src/utils/code-import.ts

```diff
@@ -0,0 +1,154 @@
+// remark-code-import
+// code-import.ts
+// https://github.com/kevin940726/remark-code-import
+import fs from "node:fs";
+import path from "node:path";
+import { EOL } from "node:os";
+import { visit } from "unist-util-visit";
+import stripIndent from "strip-indent";
+import type { Root, Code, Parent } from "mdast";
+import type { VFile } from "vfile";
+
+interface CodeImportOptions {
+  async?: boolean;
+  preserveTrailingNewline?: boolean;
+  removeRedundantIndentations?: boolean;
+  rootDir?: string;
+  allowImportingFromOutside?: boolean;
+}
+
+function extractLines(
+  content: string,
+  fromLine: number | undefined,
+  hasDash: boolean,
+  toLine: number | undefined,
+  preserveTrailingNewline = false,
+) {
+  const lines = content.split(EOL);
+  const start = fromLine || 1;
+  let end: number;
+  if (!hasDash) {
+    end = start;
+  } else if (toLine) {
+    end = toLine;
+  } else if (lines[lines.length - 1] === "" && !preserveTrailingNewline) {
+    end = lines.length - 1;
+  } else {
+    end = lines.length;
+  }
+  return lines.slice(start - 1, end).join("\n");
+}
+
+function codeImport(options: CodeImportOptions = {}) {
+  const rootDir = options.rootDir || process.cwd();
+
+  if (!path.isAbsolute(rootDir)) {
+    throw new Error(`"rootDir" has to be an absolute path`);
+  }
+
+  return function transformer(tree: Root, file: VFile) {
+    const codes: [Code, number | null, Parent][] = [];
+    const promises: Promise<void>[] = [];
+
+    visit(tree, "code", (node, index, parent) => {
+      codes.push([node as Code, index as null | number, parent as Parent]);
+    });
+
+    for (const [node] of codes) {
+      const fileMeta = (node.meta || "")
+        .split(/(?<!\\) /g)
+        .find(meta => meta.startsWith("file="));
+
+      if (!fileMeta) {
+        continue;
+      }
+
+      // if (!file.dirname) {
+      //   throw new Error('"file" should be an instance of VFile');
+      // }
+
+      // regex to get group of code from line number to line number
+      const res =
+        /^file=(?<path>.+?)(?:(?:#(?:L(?<from>\d+)(?<dash>-)?)?)(?:L(?<to>\d+))?)?$/.exec(
```

Comment:
we should also support wrapping the `file` value in quotes, both single and double:
```suggestion
        /^file=(['"])?(?<path>.+?)(?:(?:#(?:L(?<from>\d+)(?<dash>-)?)?)(?:L(?<to>\d+))?)?\1$/.exec(
```